File: server/jobs/orchestrator.js
// The Orchestrator: admission (capacity/resource gating, generic-kind tool
// set for now — research/files/computer arrive in Phase 4), steady-state
// supervision (starting queued jobs, hang detection), and crash recovery
// (classifying anything found `running` at startup). Reasoning happens only
// at the rare boundaries the plan calls for (none yet in Phase 2 — the
// admission-planning model call arrives in Phase 3, once there's a live
// conversation to announce the decomposition to); everything here is
// deterministic code, per turn.
//
// Two channels, on purpose, same shape as server/monitor/engine.js's
// monitorEvents: worker.js (and job-actions.js) emit on jobEvents
// (in-process only, so neither needs to import events.js) and THIS file
// re-broadcasts to the browser via events.js — only server.js imports this
// module, so nothing under server/tools/ ever reaches it (root CLAUDE.md's
// circular-import invariant: this file imports worker.js, which imports
// models/runner.js, which imports capabilities.js, which imports
// tools/index.js — a tool importing THIS file would deadlock the very
// loader that's trying to load it). Tool-safe job actions live in
// job-actions.js instead, a leaf module this file re-exports
// createJobIfCapacity/cancelJob/resumeStuckJob from unchanged, so
// server.js's routes don't need to know the split exists.

import * as jobStore from './job-store.js';
import { jobEvents } from './job-events.js';
import { driveJob } from './worker.js';
import { classifyRecovery, resourceAvailable, isHung, canAutoRetry, hasCapacity } from './job-policy.js';
import { resetSession, loadSnapshot } from '../conversation.js';
import { listCapabilities } from '../capabilities.js';
import { broadcast } from '../events.js';
import { getPrefs } from '../prefs.js';
import { addNotification } from '../notifications.js';

function notify(job, level, title, body) {
  try {
    addNotification({ kind: 'job', level, title, body, meta: { jobId: job.id } });
  } catch (err) {
    console.error('[jobs] notification failed:', err);
  }
}

export { createJobIfCapacity, cancelJob, resumeStuckJob, hasBackgroundCapacityNow, runningSummaries } from './job-actions.js';

const TICK_MS = 5_000;
// Generous relative to real model latency on a slow free-tier provider —
// background jobs aren't latency-sensitive the way live voice is (compare
// control/session.js's much tighter DECIDE_TIMEOUT_MS = 45s for a live,
// watched desktop-control loop). This only needs to catch a GENUINE hang
// (heartbeat silence, not "thinking hard") — see job-policy.js's isHung().
const HANG_TIMEOUT_MS = 90_000;

let tickHandle = null;
const activeWorkers = new Set(); // job ids this PROCESS is currently driving

function nowIso() {
  return new Date().toISOString();
}

// research/files restrict the generic runTurn loop to a curated tool list —
// the model still reasons iteratively (that's the whole point of a
// tool-calling loop over a single canned call into research.js/run_code.js
// directly: a real background research job often needs several searches,
// cross-referencing, deciding when it's actually answered the question).
// 'computer' is the one genuinely different kind — it bypasses the
// tool-calling loop entirely and drives control/session.js's own
// perceive/decide/act engine instead (see worker.js's driveComputerJob) —
// so it needs no tool list here at all.
const KIND_TOOL_NAMES = {
  research: ['web_search', 'read_web_page', 'look_it_up', 'check_claim', 'get_headlines', 'search_conversations'],
  files: ['run_code', 'analyze_spreadsheet', 'allow_folder'],
};

// Every runTurn-based kind gets these three, regardless of what's otherwise
// restricted — the completion/stuck/split signaling vocabulary is the same
// for every kind, not something 'research'/'files' should have to give up
// in exchange for a narrower tool list.
const INTERNAL_JOB_TOOLS = ['report_job_done', 'report_job_stuck', 'request_job_split'];

/**
 * The tool set + name->kind map for one runTurn-based job kind
 * ('generic'/'research'/'files' — 'computer' never calls this, see
 * startWorker below). Recomputed per job rather than cached: a connector's
 * tools can change between jobs (connected/disconnected mid-use).
 */
function buildToolsetForKind(kind) {
  const caps = listCapabilities({ includeMeta: false }); // already excludes internal tools
  const filtered = kind === 'generic' ? caps : caps.filter((c) => KIND_TOOL_NAMES[kind]?.includes(c.name));
  if (kind !== 'generic' && !KIND_TOOL_NAMES[kind]) {
    throw new Error(`Unknown job kind "${kind}".`);
  }
  const kindByName = new Map(filtered.map((c) => [c.name, c.kind]));
  for (const name of INTERNAL_JOB_TOOLS) kindByName.set(name, 'builtin');
  const allowedTools = [...filtered.map((c) => c.name), ...INTERNAL_JOB_TOOLS];
  return { allowedTools, kindByName };
}

function startWorker(jobId, { resumeText } = {}) {
  if (activeWorkers.has(jobId)) return;
  activeWorkers.add(jobId);
  const job = jobStore.getJob(jobId);
  // 'computer' bypasses the tool-list machinery entirely — driveJob()
  // itself branches on job.kind and never reads allowedTools/kindByName for
  // that kind.
  const toolset = job.kind === 'computer' ? {} : buildToolsetForKind(job.kind);
  driveJob(jobId, { ...toolset, resumeText })
    .catch((err) => {
      console.error(`[jobs] worker for "${jobId}" threw:`, err);
      jobStore.updateJob(jobId, { status: 'failed', error: err?.message || 'Something went wrong.', finishedAt: nowIso() });
      jobEvents.emit('status', { jobId, status: 'failed', title: job?.title });
    })
    .finally(() => activeWorkers.delete(jobId));
}

/**
 * Resuming/restarting a parked job goes through the SAME capacity gate a
 * fresh job does — without this, resuming several parked jobs in a row
 * could silently exceed prefs.maxBackgroundJobs through a back door
 * `work_in_background` itself would never allow.
 */
function assertCapacityForResume() {
  const active = jobStore.listActiveJobs();
  const prefs = getPrefs();
  if (!hasCapacity(active, prefs.maxBackgroundJobs)) {
    throw new Error(
      `Already at capacity (${active.length} of ${prefs.maxBackgroundJobs} background jobs running) — stop one or wait for one to finish first.`
    );
  }
}

/**
 * A job that's gone truly silent (no heartbeat at all — see isHung()) can't
 * be safely awaited forever, but its in-flight runTurn call also can't be
 * cancelled (no cancellation token exists in the adapters — same accepted
 * limitation as control/session.js's raceAgainstStop). Rather than trust
 * whatever partial state the hung call may have left in conversation.js's
 * live session, this discards it and resumes from the last COMPLETED step's
 * transcript snapshot — the abandoned call, if it ever resolves, finds the
 * job already moved on and its result is simply moot.
 */
async function recoverFromHang(job) {
  if (!canAutoRetry(job)) {
    jobStore.appendTrace(job.id, {
      phase: 'outcome',
      effect: 'read',
      kind: 'decision',
      summary: 'Stopped responding for a long time, and a retry already happened once — escalating instead of trying again.',
    });
    jobStore.updateJob(job.id, { status: 'awaiting_decision' });
    jobEvents.emit('status', { jobId: job.id, status: 'awaiting_decision', title: job.title });
    jobStore.addOutboxEntry(job.id, {
      tier: 1,
      reason: 'stuck',
      summary: `"${job.title}" stopped responding and a retry didn't help. Keep trying, or stop here?`,
    });
    notify(job, 'warning', `"${job.title}" needs your input`, 'Stopped responding and a retry did not help.');
    activeWorkers.delete(job.id);
    return;
  }

  jobStore.appendTrace(job.id, {
    phase: 'outcome',
    effect: 'read',
    kind: 'decision',
    summary: 'Stopped responding for a long time — resuming from the last saved point.',
  });
  jobStore.updateJob(job.id, { retries: job.retries + 1 });
  jobStore.touchHeartbeat(job.id); // buys time before the next tick re-triggers this on the same silence

  // resetSession/loadSnapshot are no-ops for a 'computer' job (it has no
  // conversation.js session at all — control/session.js manages its own
  // state) — harmless to call regardless of kind.
  resetSession(`job:${job.id}`);
  loadSnapshot(`job:${job.id}`, job.transcript || []);

  const toolset = job.kind === 'computer' ? {} : buildToolsetForKind(job.kind);
  // Deliberately bypasses the activeWorkers guard's usual add() — it's
  // already held from the original (now-abandoned) call; this just starts a
  // fresh driveJob() under the SAME jobId slot.
  driveJob(job.id, toolset)
    .catch((err) => {
      console.error(`[jobs] worker for "${job.id}" threw after hang recovery:`, err);
      jobStore.updateJob(job.id, { status: 'failed', error: err?.message || 'Something went wrong.', finishedAt: nowIso() });
      jobEvents.emit('status', { jobId: job.id, status: 'failed', title: job.title });
    })
    .finally(() => activeWorkers.delete(job.id));
}

async function tick() {
  // Start anything queued and not already resource-blocked. Capacity itself
  // was already checked at CREATION time (see createJobIfCapacity below) —
  // "never silently queue" means the caller is told up front, not that the
  // tick quietly holds a row back once it exists. A job can still sit
  // briefly `queued` here only for genuine resource contention (e.g. two
  // jobs both wanting the exclusive 'computer' claim).
  for (const job of jobStore.listJobs({ status: 'queued' })) {
    if (activeWorkers.has(job.id)) continue;
    if (!resourceAvailable(jobStore.listResourceHoldingJobs(), job.resource)) continue;
    // `resumeNote` (job-actions.js's resumeStuckJob, itself leaf-safe for
    // server/tools/ to call) is how a tool's "keep trying" response reaches
    // the worker without that tool importing this file — cleared the
    // moment it's read so a LATER re-queue of the same job never reuses
    // stale guidance.
    const resumeText = job.resumeNote || undefined;
    if (job.resumeNote) jobStore.updateJob(job.id, { resumeNote: null });
    startWorker(job.id, { resumeText });
  }

  const now = Date.now();
  for (const job of jobStore.listJobs({ status: 'running' })) {
    if (!activeWorkers.has(job.id)) continue; // not ours to supervise this tick (shouldn't happen outside a crash, which recoverOrphans() handles separately at startup)
    if (isHung(job, { now, timeoutMs: HANG_TIMEOUT_MS })) {
      recoverFromHang(job).catch((err) => console.error(`[jobs] hang recovery for "${job.id}" failed:`, err));
    }
  }
}

function describeOrphan(job, recovery) {
  const labels = {
    resumable: "It hadn't done anything I can't safely pick back up — I can resume it right where it left off.",
    restartable: "It had started some local work I can't fully trust after the restart, but nothing outside its own workspace — safe to start over.",
    needs_input: 'It was already waiting on a decision from you when this happened.',
    unrecoverable: "It had already taken an action I can't safely repeat or undo — I won't restart or resume this one automatically.",
  };
  return `While I was restarting, "${job.title}" was still marked as running. ${labels[recovery] || ''} Resume, restart, or discard it?`;
}

/**
 * Any job still `status: 'running'` at startup crashed — no process is
 * running it, so no heuristic is needed to know that (see job-policy.js's
 * classifyRecovery doc comment). Classified from its write-ahead trace,
 * marked `orphaned` (never auto-resumed), and reported via a Tier 1 outbox
 * row — the owner chooses resume/restart/discard, exactly per the build
 * spec ("never tell me it can resume if it can't"). Run once, at startup,
 * before the first tick.
 */
export function recoverOrphans() {
  for (const job of jobStore.listJobs({ status: 'running' })) {
    const trace = jobStore.getTrace(job.id);
    const recovery = classifyRecovery(job, trace);
    jobStore.updateJob(job.id, { status: 'orphaned', recovery });
    jobStore.appendTrace(job.id, {
      phase: 'outcome',
      effect: 'read',
      kind: 'decision',
      summary: `Found running with no active worker after a restart — classified "${recovery}".`,
    });
    jobStore.addOutboxEntry(job.id, { tier: 1, reason: 'permission', summary: describeOrphan(job, recovery) });
    notify(job, 'warning', `"${job.title}" was interrupted`, describeOrphan(job, recovery));
  }
}

/** Resumes an orphaned job classified `resumable` — rehydrates its transcript snapshot and continues, per job-store.js's transcript column and worker.js's own resume check. */
export function resumeOrphan(jobId) {
  const job = jobStore.getJob(jobId);
  if (!job || job.status !== 'orphaned') throw new Error('That job is not waiting on a resume/restart/discard decision.');
  if (job.recovery !== 'resumable') throw new Error(`This job is classified "${job.recovery}", not resumable — restart or discard it instead.`);
  assertCapacityForResume();
  jobStore.updateJob(jobId, { recovery: null });
  startWorker(jobId);
}

/** Starts an orphaned job over from scratch — disallowed for `unrecoverable` (restarting risks repeating an action that can't be safely repeated). */
export function restartOrphan(jobId) {
  const job = jobStore.getJob(jobId);
  if (!job || job.status !== 'orphaned') throw new Error('That job is not waiting on a resume/restart/discard decision.');
  if (job.recovery === 'unrecoverable') {
    throw new Error("This job already did something that can't be safely repeated — it can only be discarded, not restarted.");
  }
  assertCapacityForResume();
  jobStore.updateJob(jobId, { transcript: null, retries: 0, recovery: null });
  resetSession(`job:${jobId}`);
  startWorker(jobId);
}

/** Starts the 5-second tick and runs the one-time startup orphan sweep. Call once from server.js, after the HTTP server is listening — same convention as scheduler.js's startScheduler(). */
export function startOrchestrator() {
  if (tickHandle) return;
  jobEvents.on('status', ({ jobId, status, title }) => {
    broadcast({ type: 'job_progress', jobId, status, title });
  });
  recoverOrphans();
  tick().catch((err) => console.error('[jobs] initial tick failed:', err));
  tickHandle = setInterval(() => {
    tick().catch((err) => console.error('[jobs] tick failed:', err));
  }, TICK_MS);
}

File: backend/jarvis/jobs/orchestrator.py
"""Admission, supervision, and the one automatic retry.

The Conversation Manager (a live turn) decides WHETHER to background something
and does all the talking. This decides HOW the work actually gets done: whether
there is room for it, whether it needs a resource someone else holds, whether a
running job has stopped making progress, and whether a failure is worth one more
attempt or has to go to the user.

**Capacity is checked BEFORE anything is spent on planning**, so a full queue
does not cost a model call to discover. And a job is never silently queued past
capacity: the user is told and asked what should give way, because a queue nobody
can see is worse than a refusal they can answer.

**One automatic retry, total.** A crash retry and a stall retry share the same
counter, so a job cannot get two attempts by failing in two different ways.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any

from ..events import EventType, bus as default_bus
from ..events.bus import EventBus
from ..jscompat import now_iso
from . import job_store, worker
from .policy import (
    DIAGNOSE_TAIL_SIZE, can_auto_retry, classify_recovery, diagnose_stall, has_capacity,
    is_hung, resource_available,
)

logger = logging.getLogger(__name__)

#: How many jobs may be in flight at once. Small on purpose: these compete for
#: the same rate-limited model roster as the conversation the user is having.
MAX_ACTIVE_JOBS = 3

#: No heartbeat for this long on a running job means the process is gone.
HANG_TIMEOUT_MS = 5 * 60 * 1000

#: Kinds that hold something only one job may hold at a time.
RESOURCE_BY_KIND = {"computer": "computer"}


class AtCapacity(RuntimeError):
    def __init__(self, active: list[dict[str, Any]]):
        super().__init__("Already working on as much as I can at once.")
        self.active = active


def admit(*, title: str, goal: str, kind: str = "generic", priority: int = 2,
          conversation_id: str | None = None, parent_id: str | None = None,
          event_bus: EventBus | None = None) -> dict[str, Any]:
    """Create a job if there is room for it, and start it unless it must not start."""
    ebus = event_bus or default_bus
    active = job_store.list_active_jobs()
    if not has_capacity(active, MAX_ACTIVE_JOBS):
        raise AtCapacity(active)

    resource = RESOURCE_BY_KIND.get(kind)
    if not resource_available(active, resource):
        raise AtCapacity([j for j in active if j.get("resource") == resource])

    # A job that operates the real desktop never starts unattended: that is
    # exactly the outward-facing, hard-to-undo work that has to come back to the
    # owner first. It is parked awaiting a decision rather than queued, and the
    # ordinary "keep going" is the only thing that ever starts it.
    starts_parked = kind == "computer"
    job = job_store.create_job(
        title=title, goal=goal, kind=kind, priority=priority,
        conversation_id=conversation_id, parent_id=parent_id, resource=resource,
        status="awaiting_decision" if starts_parked else "queued")

    if starts_parked:
        job_store.add_outbox(tier=1, job_id=job["id"], reason="permission",
                             summary=f'"{title}" would take over the computer. OK to start?')
        ebus.publish(EventType.JOB_CREATED, {"id": job["id"], "status": job["status"]})
        return job

    ebus.publish(EventType.JOB_CREATED, {"id": job["id"], "title": title, "kind": kind})
    worker.run_in_background(job["id"], event_bus=ebus)
    return job


def resume(job_id: str, *, guidance: str | None = None,
           event_bus: EventBus | None = None) -> dict[str, Any] | None:
    """The one way a parked job starts again — used for every kind of park, so
    there is no separate machinery for "OK to start?" versus "OK to send it?"."""
    job = job_store.get_job(job_id)
    if job is None:
        return None
    if guidance:
        job_store.append_trace(job_id, phase="intent", effect="read", kind="note",
                               summary="guidance from the user", detail=guidance)
    job_store.update_job(job_id, {"status": "queued", "error": None})
    # Only the action that actually resolves the decision marks the row
    # delivered — showing it is not resolving it.
    job_store.deliver_all_for_job(job_id)
    worker.run_in_background(job_id, event_bus=event_bus)
    return job_store.get_job(job_id)


def cancel(job_id: str, event_bus: EventBus | None = None) -> dict[str, Any] | None:
    job = job_store.update_job(job_id, {"status": "cancelled", "finishedAt": now_iso(),
                                        "currentStep": None})
    job_store.deliver_all_for_job(job_id)
    (event_bus or default_bus).publish(EventType.JOB_UPDATED,
                                       {"id": job_id, "status": "cancelled"})
    return job


def supervise(now_ms: float | None = None, event_bus: EventBus | None = None) -> list[dict[str, Any]]:
    """One pass over everything in flight. Returns what it did, for the tests and
    for a real "what happened while I was away" answer."""
    ebus = event_bus or default_bus
    now_ms = now_ms if now_ms is not None else datetime.now(timezone.utc).timestamp() * 1000
    actions: list[dict[str, Any]] = []

    for job in job_store.list_active_jobs():
        if job["status"] == "running" and is_hung(job, now_ms=now_ms, timeout_ms=HANG_TIMEOUT_MS):
            actions.append(_recover(job, "no heartbeat — the work looks hung", ebus))
            continue
        if job["status"] == "running":
            stall = diagnose_stall(job_store.get_trace_tail(job["id"], DIAGNOSE_TAIL_SIZE))
            if stall is not None:
                actions.append(_recover(job, stall["detail"], ebus, cause=stall["cause"]))

    for job in job_store.list_jobs(status="stalled"):
        actions.append(_recover(job, job.get("error") or "it stopped making progress", ebus))
    return [a for a in actions if a]


def _recover(job: dict[str, Any], detail: str, ebus: EventBus,
             cause: str = "stalled") -> dict[str, Any]:
    """Spend the single automatic retry, or hand it to the user.

    The nudge goes into the SAME session, so nothing about pausing and resuming
    discards the context the work has already built up.
    """
    job_id = job["id"]
    if can_auto_retry(job):
        job_store.update_job(job_id, {"retries": 1, "status": "queued", "error": None})
        job_store.append_trace(job_id, phase="intent", effect="read", kind="note",
                               summary=f"retrying after {cause}", detail=detail)
        worker.run_in_background(job_id, event_bus=ebus)
        return {"job": job_id, "action": "retried", "cause": cause}

    job_store.update_job(job_id, {"status": "awaiting_decision",
                                  "error": detail, "currentStep": None})
    job_store.add_outbox(tier=1, job_id=job_id, reason="stuck",
                         summary=f'"{job["title"]}" is stuck: {detail}')
    ebus.publish(EventType.JOB_UPDATED, {"id": job_id, "status": "awaiting_decision",
                                         "error": detail})
    return {"job": job_id, "action": "escalated", "cause": cause}


def recover_orphans(event_bus: EventBus | None = None) -> list[dict[str, Any]]:
    """At startup: a job marked `running` with nothing running it crashed.

    No heuristic is needed to know that — the process is gone. What the trace
    decides is whether picking it back up is SAFE, and the verdict is read from
    the record rather than asked of the job.
    """
    ebus = event_bus or default_bus
    out: list[dict[str, Any]] = []
    for job in job_store.list_jobs(status="running"):
        verdict = classify_recovery(job, job_store.get_trace(job["id"]))
        job_store.update_job(job["id"], {"recovery": verdict})
        if verdict == "resumable":
            job_store.update_job(job["id"], {"status": "queued"})
            worker.run_in_background(job["id"], event_bus=ebus)
        elif verdict == "restartable":
            job_store.update_job(job["id"], {"status": "queued"})
            worker.run_in_background(job["id"], event_bus=ebus)
        else:
            job_store.update_job(job["id"], {"status": "awaiting_decision"})
            job_store.add_outbox(
                tier=1, job_id=job["id"], reason="crashed",
                summary=(f'"{job["title"]}" stopped part-way through something that '
                         "cannot safely be repeated. What would you like to do?"
                         if verdict == "unrecoverable"
                         else f'"{job["title"]}" is waiting on you.'))
        out.append({"job": job["id"], "recovery": verdict})
    return out
